### PR TITLE
Steam Controller 2 (2026) driver, Steam Deck support, multi-controller UX, and OSD/GUI fixes

### LINKED ISSUE #98: Steam Controller (2026) support
With the new steam controller being released May 4th, will this program work with the new steam controller on linux?

---

(Martin): I'll edit in the TODOs here for visibility:

- [ ] Scrolling in the OSD with SC2 is way too fast
- [x] https://github.com/C0rn3j/sc-controller/commit/16c79284c84ac93e63def76a7a54f34c0c0c22d9 - Turning SC2 off from OSD menu works but SCC does not seem to register the controller being gone
- [ ] Touchpad input mapper isn't tilted like the touchpads themselves are, so it looks a bit off
- [x] RSTICKPRESS has the wrong area defined (bigger ring instead of inner)
- [x] https://github.com/C0rn3j/sc-controller/commit/29466f212b9b43951e98b97209037139ee70a6e1 - There is no haptic feedback when pressing down on each trackpad, which is important to have on this specific controller, as there is no physical button - NOTE: how does Deck handle this?
- [ ] It would be nice to be able to disable Steam Controller support in the settings for use with Steam, and also have separate toggles for the V1 and V2
- [x] https://github.com/C0rn3j/sc-controller/commit/a410e2657acf9cbc14c90a75ffee011c42ee6e5f - RSTICK seems to be severely more limited than LSTICK when it comes to hitting corners (right upper corner on RSTICK is not hittable it seems - hard to test)

### PR DESCRIPTION
**Pull request by Patola**
**Fixes #98** — yes, it works with the new Steam Controller. 🙂

## Summary

This branch adds a full user-mode driver for the new **Steam Controller 2 (2026)**, substantially improves **Steam Deck** support, introduces **multi-controller** UX (a controller selector plus per-controller profiles, images and icons), and fixes a range of **Input-Test**, **OSD**, **GUI/mapper** and **packaging** issues. It closes with an asset-optimization (**svgo**) and **type-annotation** pass.

Everything here has been **hardware-tested** on real devices — Steam Controller v1 (wired and dongle), the Steam Controller 2 wireless Puck, and the Steam Deck (via AppImage). Commits are kept **focused (one thing each)** for bisectability. This is an LLM-assisted contribution; the human author drove the hardware reverse-engineering and testing.

---

## Steam Controller 2 (2026) driver — `scc/drivers/sc2.py`

- Reverse-engineered the v2 HID protocol (report `0x42`) from USB captures. Protocol notes and a capture/diff harness are included under `docs/` and `tools/sc2-probe/`.
- New driver for the wireless **"Controller Puck"** (`0x1304`, a 4-slot dongle) and the **wired USB-C** transport (`0x1302`); Bluetooth (`0x1303`) is scaffolded as a follow-up.
- Full input decode: buttons, both sticks, a real D-pad, both trackpads (position + pressure + **capacitive touch**), analog and digital triggers, capacitive **stick-touch** and **handle-grip** sensors, and the **IMU** (accel / quaternion / gyro — polarity verified live).
- Click **haptics** (output report `0x82`); lizard-mode kept disabled; controllers survive turn-off/on (the Puck re-attaches without tearing down the dongle).
- GUI: a dedicated v2 controller image + `images/sc2.config.json`. The capacitive stick-touch and grip sensors are bindable directly, or usable as conditions in mode-shift combinations.

## Steam Deck

- Capacitive stick-touch parity with the v2.
- Correct button layout (D-Pad to the top, Steam to the left column), back-button mapping, and rear-paddle order (**L4/R4 above L5/R5**, matching the device).
- Per-controller icons under `images/deck/`.
- Defaults "disable emulation on close" **on** (the Deck has no built-in keyboard, so a lingering daemon keeps emulating).

## Multi-controller UX

- A **controller selector** replaces the per-controller profile bars: pick a controller, then its profile. The first-connected controller is primary.
- Each controller **remembers its own profile** across (re)connects — by connection slot, or by hardware serial when "Use Serial Numbers to Identify Controllers" is on.
- **Per-controller button-image overrides** and distinct controller **icons** for v1 / v2 / Deck.
- New README section documenting multi-controller usage (with a screenshot).

## Input Test mode

- Works for the v2 and the Deck; shows the **right stick and D-pad**; fixes highlight/test-area offsets; a viewBox-aware cursor; re-arms on reconnect; and clears highlights when the mode is turned off.

## OSD

- Menu fixes: **Wayland** window shaping (a cairo circle-clip instead of the X SHAPE extension), **Python 3.13** compatibility, long-menu scrolling, and autoswitch handling.
- A **per-controller binding display** with a dedicated v2 layout.
- On-screen keyboard fixed on Wayland; "Edit Bindings" opens the OSD-keyboard editor; removed the abandoned controller-driven OSD edit mode.

## GUI / mapper

- An **"Act on release"** (inverted-button) modifier, useful for always-on sensors like grip touch.
- All four grips + right-stick press offered in the button and mode-shift choosers.
- The Steam logo on the SC v1 "C" button, sized to match the generic button.
- Mapper fixes for non-Deck `HAS_RSTICK` controllers (right pad / stick / D-pad) and a v2 right-stick crash under a mode modifier.

## v1 / general fixes

- Repair a Python 2→3 port bug that dropped v1 controllers on hotplug; keep the v1 dongle alive when `GET_SERIAL` stalls; spawn helpers with a valid interpreter when `sys.executable` is bogus (AppImage); build the `uinput` enums via the functional `IntEnum` API; and repair "register new controller" and "Restart emulation".

## Packaging / AppImage

- Bundle **libxml2 + ICU** (so librsvg works on minimal hosts) and **ayatana-appindicator + libdbusmenu** (so the tray icon works); brand the AppImages as `sc-controller-cc`; ship a single, most-compatible **jammy** build; and document the one-time udev-rules step (noting the Steam Deck doesn't need it).

## Asset optimization (svgo)

The 50 shipped SC1/SC2/Deck SVGs (plus 7 source assets) are run through `svgo` — **~30% smaller** with no visible change. These SVGs aren't only artwork; the GTK GUI reads them in ways a naive minifier silently breaks, so `tools/svgo.config.js` disables exactly the passes that would:

- **`cleanupIds` off** — the GUI looks elements up by id (`AREA_*` regions, `controller`/`root`/`background`/`label_template`, per-control ids).
- **`convertShapeToPath` off** — hover / Input-Test areas are read as `<rect x/y/width/height>`.
- **`removeViewBox` / `removeHiddenElems` off** — coordinate math needs the viewBox; the Input-Test overlay lives in a `display:none` layer.
- **`removeUnknownsAndDefaults` off** — custom attributes (e.g. `scc-button-scale`) must survive.
- **`collapseGroups` / `moveElemsAttrsToGroup` / `moveGroupAttrsToElems` off** — button glyphs are `<g id="button"><g transform=…><path/></g></g>`, and `controller_image._fill_button_images` **overwrites** the `button` group's transform when placing a glyph; flattening lands `id="button"` on the path with the normalisation transform, which then gets clobbered.
- **`convertTransform` off** — `SVGEditor.parse_transform` uses a comma-only regex; svgo's `translate(a,b)` → `translate(a b)` rewrite makes the GUI read the element as untransformed, dropping the `sc`/`deck` `controller`-group offset.

`tools/gen_sc2_image.py` and `tools/gen_binding_display.py` now pipe their output through `svgo` (helper `tools/_svgo.py`) so regenerated assets stay optimized, with a source-preserving config variant for the generator-parsed art.

## Type annotations

Every function, class and parameter this branch introduces now carries type annotations (driver, GUI, OSD, mapper, `tools/`, tests) — `ruff`'s flake8-annotations (`ANN`) rules pass on the added lines. Forward references use `from __future__ import annotations` + `TYPE_CHECKING`, matching the existing style. No runtime behaviour change.

---

## Testing

- **Hardware:** SC v1 (wired + dongle), SC v2 Puck, and the Steam Deck (AppImage build).
- **svgo:** config verified to preserve every element id, the `<g id="button">` structure, comma-separated transforms and `AREA` geometry byte-for-byte; all SVGs render (`rsvg-convert`); the generators still produce valid output.
- **Typing:** `ruff --select ANN` clean on the added lines; unit tests pass; full import of the touched modules succeeds.

### CODE REVIEW

File: .github/workflows/appimage.yml [hunk 1]
```diff
@@ -16,26 +16,14 @@ jobs:
...
-            codename: bookworm
-            pubkey: 54404762BBB6E853
-            pubkey_security: 6ED0E7B82643E131
-            recipe: AppImageBuilder.debian.yml
-          - os: debian
-            codename: trixie
-            pubkey: BDE6D2B9216EC7A8
-            pubkey_security: 8E9F831205B4BA95
-            pubkey_updates: 78DBA3BC47EF2265
-            recipe: AppImageBuilder.debian.yml
```
Comment (C0rn3j) [inline .github/workflows/appimage.yml]: Why drop it?

Now we won't know when something breaks on other distros/versions, and we'll be in for a rude awakening when bumping the version.

Dropping the checksums is fine, I don't think they serve much of a purpose, but do keep the builds for other OSs.

File: AppImageBuilder.yml [hunk 2]
```diff
@@ -200,4 +207,4 @@ AppDir:
 
 AppImage:
   arch: "{{APPIMAGE_ARCH}}"
-  update-information: "gh-releases-zsync|C0rn3j|sc-controller|latest|sc-controller-*-{{APPIMAGE_APT_DISTRO}}-{{APPIMAGE_ARCH}}.AppImage.zsync"
+  update-information: "gh-releases-zsync|Patola|sc-controller-cc|latest|sc-controller-cc-*-{{APPIMAGE_APT_DISTRO}}-{{APPIMAGE_ARCH}}.AppImage.zsync"
```
Comment (C0rn3j) [inline AppImageBuilder.yml:210]: What's the point of renaming to `sc-controller-cc`?

Or changing the author really.

File: scc/drivers/ds4drv.py [hunk 3]
```diff
@@ -408,6 +419,187 @@ def _generate_id(self) -> str:
 		return id
 
 
+# --- Bluetooth hidraw driver ----------------------------------------------------
```
Comment (C0rn3j) [inline scc/drivers/ds4drv.py:522]: Anything wrong with the PR for DS4 hidraw that's currently open (due to me unfortunately somewhat ignoring it)?

File: AppImageBuilder.yml [hunk 4]
```diff
@@ -126,15 +130,18 @@ AppDir:
...
+      # them matches what the older appimage-build.sh shipped.
       - libstdc*        # development
-      - libxml*         # codec
 
       # libgtk3-0
       - "*-icon-theme"  # gui
       - "*crypt*"       # security
       - "*dconf*"       # gui configuration
-      - "*dbus*"        # process communication
+      - "libdbus-1*"    # process communication (keep libdbusmenu for the tray icon)
```
Comment (C0rn3j) [inline AppImageBuilder.yml:144]: the previous `*dbus*` already both matched the `libdbus-1` and `libdbusmenu-*`, what's the improvement with singling them out like this instead?

**Comment by C0rn3j** [comment]
Awesome work getting things working and fixed up.

As I hinted at in https://github.com/C0rn3j/sc-controller/issues/98#issuecomment-4844655764, please take everything that's not exclusively related to SC2 changes, and ship it as separate PRs, per problem.

That lets us start reducing the changeset in a fashion that enables proper reviews.

File: scc/uinput.py [hunk 5]
```diff
@@ -49,28 +49,24 @@
...
-
-class Rels(IntEnum):
-	"""Rels enum contains all rels from linux/uinput.h (REL_*)."""
-
-	locals().update({i: CHEAD[i] for i in CHEAD if i.startswith("REL_")})
+# Build these enums from the parsed kernel constants in CHEAD via the functional
+# IntEnum(...) API. A class body that instead did `locals().update(...)` relies
+# on dict.update() reaching enum's _EnumDict.__setitem__ -- which it does not, so
+# member registration depends on CPython's enum internals: it happens to work on
+# the newer Python we develop on but silently yields an EMPTY enum on Python 3.10
```
Comment (C0rn3j) [inline scc/uinput.py]: Let's revert this, I'll just bump the project up to 3.11 minimum, 3.10 is EOL in about 3 months, no point in trying to keep it alive with workarounds.

File: .github/workflows/appimage.yml (hunk 1 above)
Comment (Patola) [reply .github/workflows/appimage.yml]: Oversight. I did not ask the LLM to remove the builds, I just asked it to not include in the releases. I will restore them.

File: AppImageBuilder.yml (hunk 2 above)
Comment (Patola) [reply AppImageBuilder.yml:210]: I meant no disrespect. For the future of this repository, once I get all the functionality to the upstream, I intend to do wild experiments with the fork, not even worrying to much about practicality, like emulating a 3D Connexion 3D Mouse, trying to inject an "Steam Actions" sniffer in games to be able to react to steamworks inputs, and things like that. And for that I want to make it clear it was done with "claude code" (hence the "-cc").

As for the authorship change, again, LLM doing what I didn't ask for, will revert that.

File: scc/drivers/ds4drv.py (hunk 3 above)
Comment (Patola) [reply scc/drivers/ds4drv.py:522]: Because the existing DS5 hidraw code was almost the same we needed for DS4 and it was true and tested.

File: AppImageBuilder.yml (hunk 2 above)
Comment (Patola) [reply AppImageBuilder.yml:210]: I have also changed the name on the appimages and packages as to not try and pass for the official packages, and pointed to my repository provisionally to not pass as the "official" sc-controller in any fashion.

File: AppImageBuilder.yml (hunk 4 above)
Comment (Patola) [reply AppImageBuilder.yml:144]: Claude Code's answer:

The key point: this is an exclude list
include/exclude are apt package-name lists, and exclude globs are matched against package names and prune them from the bundle (so those libs come from the host instead). *dbus* matches both:

libdbus-1-3 — the low-level D-Bus IPC lib
libdbusmenu-glib4 / libdbusmenu-gtk3-4 — the SNI/Ayatana tray-menu libs, which are explicitly in include: (lines 111–112)
So "already matched both" is exactly the bug, not a convenience. In an exclude list, matching libdbusmenu-* means stripping it back out of the bundle — directly undoing the include: we added for it. On hosts that ship libdbusmenu it happened to still work; on minimal hosts that don't (SteamOS / Steam Deck) the tray/status icon menu was dead, because the one copy we intended to bundle got excluded.

Why libdbus-1* specifically
We do still want to drop libdbus-1 — it's universally present on every target host, and unbundling it (using the host's) is the standard way to avoid an AppImage-vs-host libdbus/session-bus version mismatch. But we must keep libdbusmenu-* bundled, because minimal hosts don't provide it and it's what the tray needs.

Narrowing the glob is the fix: libdbus-1* excludes only the core IPC lib (safe to take from the host) while leaving the explicitly-included libdbusmenu-* in the bundle. The broader *dbus* couldn't express "drop libdbus-1 but keep libdbusmenu" — it collapsed both into one decision and picked the wrong one for the tray.

One-line version: *dbus* over-matched and excluded libdbusmenu-* (cancelling its include), which broke the tray icon on hosts that don't ship libdbusmenu; libdbus-1* excludes only the host-provided IPC lib and preserves the bundled tray-menu libs.

**Comment by Patola** [comment]
Added a new commit with the reverts requested by upstream. I am out of tokens now, will resume working on it (the gyroscope/accelerometer for the DS4, the Steam Controller 2 and Steam Deck) on July 9.

File: AppImageBuilder.yml (hunk 4 above)
Comment (C0rn3j) [reply AppImageBuilder.yml:144]: I have repeatedly made the mistake of not *fully* expanding the code block, so it all seemed to be in `include:` for me, I blame the bad GitHub UI/UX on this one really... welp.

**Comment by Patola** [comment]
> Awesome work getting things working and fixed up.
> 
> As I hinted at in [#98 (comment)](https://github.com/C0rn3j/sc-controller/issues/98#issuecomment-4844655764), please take everything that's not exclusively related to SC2 changes, and ship it as separate PRs, per problem.

I've been thinking about it, is that even possible now, without resetting back to sc-controller 0.5.5 and then trying to reapply the whole code with the same-ish comments? I admit that I got too enthusiastic and got carried away bu touching the entire codebase, but it was literally as I said as one of my first comments when I announced my fork -- it wasn't possible to only touch the v2 controller, there were too many things dependent on core code that I had to change.

So, the code spanned v2 progress alongside changing some core code, and a good example of that is the "Display Bindings" screen: first I went through several iterations to create a perfect binding display for the steam controller v2, only much later I realized that no controller besides the steam controller had that display bindings window... All of them would fallback to the steam controller, something that to my eyes seemed so wrong that I immediately got to generalize my Display Binding generator on the art that already existed to create something that would work for all controllers.

I had to discard my own fine-tuned Display Bindings for the controller but that was OK, since the result was worth it, but it made the web of commits very complicated, touching something specific then coming back again, rewriting that touch alongside a broad retouch of all the controllers. And there were several back-and-forths like this during my enthusiasm bursts. Trying to reduce/separate that in different PRs, besides being a pain, would certainly lose information on the history of these changes.

The code that I'm fiddling now, that is relative to the DS4 calculation of quaternions and sensor fusion, is also broad and at the same time specific; there is broken stuff both in the general parts it uses and the specific for the DS4 (that also affects the DS5 since the code is pulled from there), and I chose this path as a kind of "control group" to straighten out the general code, then get back to the steam controller 2, and fix its specifics -- and as of now I am still wary of the `0x42` / `0x45` firmware differences that I have pasted here, because if at any time the controller skips on gyroscope updates, doing the sensor fusion on the host might get too inaccurate to be useful.

I would really like to know your thoughts on this. I know that in an ideal world I should have done this code slowly and with less "mobility", but I think if I did that, that would double or triple the time it would take to land. I took the path: I see something wrong, I fix it, if not immediately, at least as soon as I got to touch its scope.

I hope that doesn't offend you (since I know how attached we can be to our code), but as a heavy user of sc-controller in the last years I have been watching features stop working -- not really fault of the code, but how things progressed, like the switch from X11 to Wayland --  and I think that an incremental growth without touching what was already there is partly responsible for the breakage, I am trying to renew some core parts as to finally re-enable the sane, slow approach of doing more controlled, well-defined PRs in the future.

My PR will still have a final code review from a friend that also uses LLM, and that will hopefully straighten up stuff even more, but it will also reinforce that this huge monolithic PR will be needed in its whole when applied, lest things fall apart. So, yes, even this as it is now is not ready for applying, I need to sort out the DS4 and SC2 gyroscope/quaternions at least and get them very well tested, plus the code review.

**Comment by C0rn3j** [comment]
> I've been thinking about it, is that even possible now, without resetting back to sc-controller 0.5.5 and then trying to reapply the whole code with the same-ish comments?

Sure.

* Reset your main branch back to upstream main branch - you shouldn't be changing the default branch in the first place, you should always work in a separate branch
*  create a new branch from main
*  cherry pick a relevant commit or two from your sc2 branch into your new branch
* send a PR from your new branch
* rinse and repeat

When it gets merged, update your main branch and rebase sc2 branch against main, which should be conflict-less since you simply cherry picked existing commits from it.

Though, that assumes your commits are atomic in the first place, which they not always are, so first resolve that:

---

In your sc2 branch, go to commits like this one - https://github.com/C0rn3j/sc-controller/commit/22418ccb8e846b378b03c2224be2b512de5e0eb7 - undo it, and save the changes as separate commits per-change instead.

Then ideally squash things together where it makes sense - i.e. reorder and squash all the appimage commits into one.

Some squashable things are already atomic, i.e. you can reorder and squash https://github.com/C0rn3j/sc-controller/commit/e06f7b6f984fa53d2bdc7dc8618a5f681d857e01 into https://github.com/C0rn3j/sc-controller/commit/3806433f6e4ab4f6caf5df1aa1bb3cb054e99919 as is.

It does not make sense to pull in 10 commits that all just edit a TODO.md, squash them into one commit, it makes for much cleaner history and is much nicer to review when going commit by commit.

If you've never done such git rebasing surgery before it is a bit daunting, but learning to do it is going to be mighty beneficial. In this case, always make sure you have a backup before you start. i.e. create a new branch `sc2-orig` from your `sc2` branch and push it online and save it somewhere aside too, so you can always restore it in case you mess up.

**Comment by Patola** [comment]
Thanks for those instructions but the issue is not technically how to do it. It is that the commits are completely dependent/entangled on one another with sometimes a few degress of indirection, a commit only to sc2 files will not work because it depends on code to other parts. As I said, I had to go into core and other subsystems to get stuff working and consistent. It is likely not possible to disentangle this.

**Comment by Patola** [comment]
I've made a reevaluation of the request of breaking by feature, and this is an attempt of an explanation of the issues (yes, done with Claude Code, revised and edited by me):

---------------------------------------------------------
I understand the concern; a 110-commit PR is a lot to review, and I want to make it easier, not harder. Here's what I can and can't realistically do:

What I'll happily split out into their own PRs (genuinely independent, not SC2-related): the packaging/CI cluster — the AppImage build-matrix/artifact changes, the libdbusmenu tray fix, the AppImage branding, and the Python-version bump (requires-python >=3.11 + the enum construction). These stand alone, build alone, and test alone. I'll rebase them onto main as separate PRs so they can land independently.

What can't be split per-problem, and why: the rest isn't "feature + cleanup" — it's _refactor-to-enable_. To get a working SC2 driver without breaking the other controllers, I had to fix crippling bugs in shared code that every controller flows through (`mapper.py`, `sccdaemon.py`, `usb.py`, `config.py`), and those fixes are causally downstream of the SC2 work:

* The mapper right-stick fixes ([af356b28](https://github.com/C0rn3j/sc-controller/pull/100/commits/af356b28d17948ab1f433f97a01fb7a8f030d1b1), [9206a4ee](https://github.com/C0rn3j/sc-controller/pull/100/commits/9206a4ee123d48a520447bd18da58a9c96e6da45), [cb9a4c05](https://github.com/C0rn3j/sc-controller/pull/100/commits/cb9a4c05eb2fedb47f5e68e215c8fe2f5e0ca751)) are three stacked layers on the same handling in mapper.py; SC2's `HAS_RSTICK` crashes the mapper without all three. In isolation each looks arbitrary — the repro and the test only exist because of the new HID driver.
* Some are legacy Python 2→3 bugs (b42597a7 dropped v1 pads on hotplug; [0e153648](https://github.com/C0rn3j/sc-controller/pull/100/commits/0e1536488e6bb823cc792822e029e2af366bfef8) Feedback bytes/str) that only surfaced under the multi-controller path.
* https://github.com/C0rn3j/sc-controller/pull/100/commits/fe2d288e860cda6359a30ea785020afdde186418 is a single type-annotation pass across four shared files at once — there's no per-problem slice of it.

Mechanically, extracting these into independent, individually-CI-green PRs means an interactive rebase reordering 110 commits with overlapping hunks in the same shared files across different "topics" — cascading conflicts, and every fragment would have to be made to build and pass tests _without the others it depends on_. That's days of high-risk history surgery on a branch that is currently working and hardware-verified on real **SC2**, **Steam Deck**, and **DS4** hardware — trading a tested whole for many untested fragments.

What I think gets you the reviewability you're after without that risk: the history is already atomic — one commit, one logical change — so it reviews cleanly commit-by-commit within a single PR. I'm happy to add a short dependency map (which fixes are prerequisites for which), and to walk through any commit's rationale. Would splitting off the infra PRs above + reviewing the SC2 core as one ordered, per-commit branch work for you?

---------------------------------------------------------
What do you think? Does this counter-proposal seems ok to you?
For the record, there are other issues and entangled changes that Claude didn't even mention, like the shared gyroscope code that might cause input to be stuck (I myself had to exit sc-controller with Alt-F4 at least half a dozen times due to this bug) and that in isolation doesn't seem to make sense -- https://github.com/C0rn3j/sc-controller/pull/100/commits/50c166b365ab13964ce7861e16c196a469756468 -- and chasing them to illustrate would be itself a lot of work. Hence I think that my proposal takes us back to the realm of feasibility, in style of that meme "that's the best I can do"...

File: scc/uinput.py (hunk 5 above)
Comment (Patola) [reply scc/uinput.py]: For the record: looked deeper into this and the `IntEnum("Keys", {...}` is not a "3.10 crutch," it's just robust code that also happens to run on the jammy AppImage. In reality reverting it re-opens their own #24 and breaks the AppImage. So I am keeping the enum there if you don't mind.

**Comment by C0rn3j** [comment]
https://github.com/C0rn3j/sc-controller/commit/b42597a7dad2ebc5f48ab2e0e4e0d7311ea6afed How do I reproduce this issue? I do have 2x SCv1 controllers.

https://github.com/C0rn3j/sc-controller/commit/387c45ca27ea4134e0c63f0411fc3e8c4373f3e1 Can be dropped.

Remaining Appimage commits can be dropped - let's keep attaching all builds, easier for users to find issues with them, we can add a nice release table with explanations instead or something if the idea was to have users prefer the oldest build.

**Comment by Patola** [comment]
> [b42597a](https://github.com/C0rn3j/sc-controller/commit/b42597a7dad2ebc5f48ab2e0e4e0d7311ea6afed) How do I reproduce this issue? I do have 2x SCv1 controllers.

Archaeology: introduced by c3dd3ad9 ("Fixed dongle connections", May 2020) during the Python 3 port — the original py2 ord(data[4]) started raising TypeError on py3 (int, not str), and wrapping it in str() silenced the exception while inverting the logic. Your current main still has it (sc_dongle.py:101).

Hardware repro (2× SCv1 + 1 wireless dongle), on current main:

1. Pair both v1s to the same dongle (one-time, via Steam if not already paired).
2. Run `scc-daemon debug` in a terminal (so you see controllers being added/removed).
3. Power on controller A, without touching any other control, and wait. The dongle sends the HOTPLUG connect packet (status `0x03`, `data[4] == 2`) — with the bug, nothing is added. An idle controller only emits IDLE (`0x04`) packets, which are ignored, so it stays invisible until some control activity produces an INPUT (`0x01`) packet (that path has its own add-fallback, which is why single-controller use mostly masked the bug).
4. Power on controller B: same story — it announces via HOTPLUG and doesn't appear.
5. The nastier half — spurious removal: any HOTPLUG connect re-announce for an endpoint that's already registered falls into the `elif endpoint in self._controllers` branch and removes the live controller. Triggers are firmware-timing dependent (re-link after a radio dropout / sleep-wake / dongle re-announce when the other controller joins), so it's intermittent — that's what made a second v1 "silently" vanish rather than fail loudly.

Expected with the fix (`data[4] == 2`): both controllers appear immediately on power-on, before any input, and re-announces are idempotent (re-add if missing, never remove).

Caveat for your test: our branch also fixes a separate v1 multi-controller issue (5c898df0, GET_SERIAL stall tearing down the dongle) — on main that one can also make a second v1 misbehave, so if you test there you may see compounded symptoms. Step 3 (power on, don't touch, watch it not appear) isolates this commit's bug cleanly.

> [387c45c](https://github.com/C0rn3j/sc-controller/commit/387c45ca27ea4134e0c63f0411fc3e8c4373f3e1) Can be dropped.

Why? This is a serious bug that caused a crash in OSD when in Wayland. Has it been fixed elsewhere?

> Remaining Appimage commits can be dropped - let's keep attaching all builds, easier for users to find issues with them, we can add a nice release table with explanations instead or something if the idea was to have users prefer the oldest build.

**Comment by C0rn3j** [comment]
> Why?

Cause it's now in master

**Comment by C0rn3j** [comment]
late edit - 

I've done almost all of the needed rebasing work myself and now keep it in the `sc2` branch - https://github.com/C0rn3j/sc-controller/compare/main...C0rn3j:sc-controller:sc2 - it would be nice if in the future you rebased your branch instead of doing merge commits to keep up with upstream, it creates more mess otherwise.

I would highly suggest to just hard reset your branch against the upstream `sc2` branch, unless you want to dupe a lot of the rebasing effort.

Note the branch drops all the DS4/DS5 commits, as main diverged too much from your fork there.  
Add any remaining necessary fixes back.

---

https://github.com/C0rn3j/sc-controller/commit/baa394e5ea2e0b9c061f382fd278211622a37203 Can be dropped.  
https://github.com/C0rn3j/sc-controller/commit/b42597a7dad2ebc5f48ab2e0e4e0d7311ea6afed Can be dropped.  
https://github.com/C0rn3j/sc-controller/commit/2b378da859e3c83b482b5482657663de133f77c5 Can be dropped.  
https://github.com/C0rn3j/sc-controller/commit/3494465fe436145d1bbd55c43e2aa650515a2ca4 Can be dropped.  
https://github.com/C0rn3j/sc-controller/commit/c86305afa5f6c7f5bc614be9979ba61ead237ca6 Can be dropped.  
https://github.com/C0rn3j/sc-controller/commit/b84d3a611baa81a5e35bd6c10f04461c595cfd4e Can be dropped.  
https://github.com/C0rn3j/sc-controller/commit/a920e429c342a8e7067dec4c711c7ae2de7059e6 Can be dropped after splitting it into last controller image (drop) and sc2 fixes (keep)
https://github.com/C0rn3j/sc-controller/commit/f40801407231cba6587bcd455abe69fe60325926 Can be dropped after splitting it into the radial fix (drop) and the locking fix (keep)
https://github.com/C0rn3j/sc-controller/commit/2ec35cedb1551889fbee7b5080323f1dc583ea34 Can be split into its five elements (sigh) and the autoswitcher changes can be dropped and the menu.py imports can be dropped also  
https://github.com/C0rn3j/sc-controller/commit/6b8c865427dfe904a5ebd29dd074e3d8158bc58a has (I think) the wrong way to fix the issue, see https://github.com/AppImage/AppImageKit/issues/1435 - if am correct, can be dropped now that I applied a workaround for appimage - and the appimage bits in https://github.com/C0rn3j/sc-controller/commit/e206727fde97fb9c3deaab0ba89af3bf8e0f9249 too can get dropped after you make it atomic

---

I can't reproduce the hotplug bug - I did once get SCC into a state where the controller was eventually dropped while leaving them idle and I had to use it for input to get it to re-appear, but could not do it twice... I'll keep trying I suppose.

EDIT: I could repro it again but your change does not seem to fix that, restarting `scc-daemon` does not pick up on the controller anyway until it is touched, so I suppose that's a different bug.

First instance log:
```
...
D SCCDaemon     Controller added: <SCWireless sc2>
D OSD DaemonCtrl    Daemon reported profile change for ds4: /home/c0rn3j/.config/scc/profiles/XBox Controller.sccprofile
D OSD DaemonCtrl    Daemon reported profile change for sc1: /home/c0rn3j/.config/scc/profiles/XBox Controller.sccprofile
D OSD DaemonCtrl    Daemon reported profile change for sc2: /home/c0rn3j/.config/scc/profiles/XBox Controller.sccprofile
I SCCDaemon     Loaded profile '/home/c0rn3j/.config/scc/profiles/XBox Controller.sccprofile'
D OSD DaemonCtrl    Daemon reported profile change for sc2: /home/c0rn3j/.config/scc/profiles/XBox Controller.sccprofile
D SCCDaemon     Controller removed: <SCWireless sc2>
```

EDIT2: I did read the history and it does indeed to seem to be wrongly converted code though, shipped the fix.

---

Running `tools/gen_sc2_image.py` generates different `sc2`, `LGRIP`, `LGRIP2`, `RGRP` and `RGRIP2` SVGs.

**Comment by syboxez** [comment]
Built both the sc2 branch in the parent repo and the steam-controller-v2 branch in your repo and noticed a few things.

When switching from the Steam Controller V2 to another controller with a smaller window, the window will shrink itself as the system cursor is moved.

There is no haptic feedback when pressing down on each trackpad, which is important to have on this specific controller, as there is no physical button.

I cannot navigate the OSD with the DPad, only with the joystick and the left trackpad.

Adjusting mouse settings on the right trackpad does nothing, including trying to adjust the sensitivity, feedback, and trackball friction. The settings still apply fine on my original Steam Controller V1.

It would be nice to be able to disable Steam Controller support in the settings for use with Steam, and also have separate toggles for the V1 and V2.

Also, is it really necessary for the V2 window size to be that much larger compared to other controllers?

**Comment by Patola** [comment]
> Built both the sc2 branch in the parent repo and the steam-controller-v2 branch in your repo and noticed a few things.
> 
> When switching from the Steam Controller V2 to another controller with a smaller window, the window will shrink itself as the system cursor is moved.
> 
> There is no haptic feedback when pressing down on each trackpad, which is important to have on this specific controller, as there is no physical button.
> 
> I cannot navigate the OSD with the DPad, only with the joystick and the left trackpad.
> 
> Adjusting mouse settings on the right trackpad does nothing, including trying to adjust the sensitivity, feedback, and trackball friction. The settings still apply fine on my original Steam Controller V1.
> 
> It would be nice to be able to disable Steam Controller support in the settings for use with Steam, and also have separate toggles for the V1 and V2.
> 
> Also, is it really necessary for the V2 window size to be that much larger compared to other controllers?

Thanks for your comments. Do you mind re-sending this as a bug report in my repository? It will be easier to track.

**Comment by C0rn3j** [comment]
I'm closing this since this wasn't mergeable initially, and isn't mergeable now that it is so far behind.

I can't even apply half the commits to the `sc2` test branch due to the incompatibilities.

If you wish to get any of the work in(without me doing all the work), I highly suggest you rebase on the sc2 branch, apply any changes you're missing on top, and go from there.

PRs focused on a single problem at a time are welcome, megaPRs like this whose contents range from small bug fixes to feature additions or GUI rework are just not able to be gotten in.

@syboxez you're welcome to report future problems in the SC2 thread for SC2 and the general tracker for the rest - I've reported/added all what you mentioned (that pertained to this repo).
